src/OmniSharp.Host/MSBuild/Discovery/MSBuildLocator.cs

```diff
@@ -0,0 +1,158 @@
+                    new DevConsoleInstanceProvider(loggerFactory),
+                    new VisualStudioInstanceProvider(loggerFactory),
+                    new MonoInstanceProvider(loggerFactory),
+                    new StandAloneInstanceProvider(loggerFactory, allowMonoPaths: true)));
+
+        public static MSBuildLocator CreateStandAlone(ILoggerFactory loggerFactory, bool allowMonoPaths)
+            => new MSBuildLocator(loggerFactory,
+                ImmutableArray.Create<MSBuildInstanceProvider>(
+                    new StandAloneInstanceProvider(loggerFactory, allowMonoPaths)));
+
+        public void RegisterInstance(MSBuildInstance instance)
+        {
+            if (_registeredInstance != null)
+            {
+                throw new InvalidOperationException("An MSBuild instance is already registered.");
+            }
+
+            _registeredInstance = instance ?? throw new ArgumentNullException(nameof(instance));
+
+            AppDomain.CurrentDomain.AssemblyResolve += Resolve;
+
+            if (instance.SetMSBuildExePathVariable)
+            {
+                var msbuildExePath = Path.Combine(instance.MSBuildPath, "MSBuild.exe");
+                var msbuildDllPath = Path.Combine(instance.MSBuildPath, "MSBuild.dll");
+
+                string msbuildPath = null;
+                if (File.Exists(msbuildExePath))
+                {
+                    msbuildPath = msbuildExePath;
+                }
+                else if (File.Exists(msbuildDllPath))
+                {
+                    msbuildPath = msbuildDllPath;
+                }
+
+                if (!string.IsNullOrEmpty(msbuildPath))
+                {
+                    Environment.SetEnvironmentVariable("MSBUILD_EXE_PATH", msbuildPath);
+                    _logger.LogInformation($"MSBUILD_EXE_PATH environment variable set to '{msbuildPath}'");
+                }
+                else
+                {
+                    _logger.LogError("Could not find MSBuild executable path.");
+                }
+            }
+
+            var builder = new StringBuilder();
+            builder.Append($"Registered MSBuild instance: {instance}");
+
+            foreach (var kvp in instance.PropertyOverrides)
+            {
+                builder.Append($"{Environment.NewLine}            {kvp.Key} = {kvp.Value}");
+            }
+
+            _logger.LogInformation(builder.ToString());
+        }
+
+        private Assembly Resolve(object sender, ResolveEventArgs e)
+        {
+            var assemblyName = new AssemblyName(e.Name);
+
+            _logger.LogDebug($"Attempting to resolve '{assemblyName}'");
+
+            if (s_msbuildAssemblies.Contains(assemblyName.Name))
+            {
+                var assemblyPath = Path.Combine(_registeredInstance.MSBuildPath, assemblyName.Name + ".dll");
+                var result = File.Exists(assemblyPath)
+                    ? Assembly.LoadFrom(assemblyPath)
```

Comment:
should this use https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Host/Services/AssemblyLoader.cs#L62 ?

Reply:
Yes, you're right. I didn't use it initially because I was still trying to figure out when/how the `IMSBuildLocator` would be initialized. New commit adds this.